utils/swift_build_support/swift_build_support/toolchain.py

```diff
@@ -0,0 +1,124 @@
+# -*- python -*-
+# swift_build_support/toolchain.py - Detect host machine's versioned executables
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+#
+# Find the path to a Clang executable on the host machine that is most
+# suitable for building Swift.
+#
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import
+
+import collections
+import platform
+import subprocess
+
+from . import xcrun
+from .which import which
+
+
+class Toolchain(object):
+    """
+    A class that stores accessible paths to system commands.
+
+    It requires a 'cc' and 'cxx' tool in its keyword args.
+    """
+    def __init__(self, cc=None, cxx=None, **kwargs):
+        self.cc = cc
+        self.cxx = cxx
+        self.tools = [cc, cxx]
+        for tool, path in kwargs.iteritems():
+            self.tools.append(path)
+            setattr(self, tool, path)
+
+
+def _freebsd_release_date():
+    """
+    Return the release date for FreeBSD operating system on this host.
+    If the release date cannot be ascertained, return None.
+    """
+    try:
+        # For details on `sysctl`, see:
+        # http://www.freebsd.org/cgi/man.cgi?sysctl(8)
+        return int(subprocess.check_output(
+            ['sysctl', '-n', 'kern.osreldate']).rstrip())
+    except subprocess.CalledProcessError:
+        return None
+
+
+def _first_common_toolchain(tools, suffixes=None):
+    """
+    Return a Toolchain of resolved paths where each path has
+    the same suffix.
+
+    If there is no common version of all binaries found, return None.
+    """
+    if suffixes is None:
+        # No suffixes provided, default to using empty suffix only
+        suffixes = ['']
+    for suffix in suffixes:
+        path_map = dict()
+        for name, tool in tools.iteritems():
+            path = which(tool + suffix)
+            if not path:
+                break
+            path_map[name] = path
+        if len(path_map) == len(tools):
+            return Toolchain(**path_map)
```

Comment:
Oh, is this why default arguments are provided above? A bit too much metaprogramming for me, but to each their own! :sweat_smile: